### PR TITLE
Update api new features

### PR DESCRIPTION
**Pull request by luguenth**
Updated to new Triggerdialog API.

You're now logged in via credentials of your Mautic User.

### CODE REVIEW

File: Config/config.php [hunk 1]
```diff
@@ -6,12 +6,14 @@
...
 use MauticPlugin\MauticTriggerdialogBundle\Form\Type\VariableType;
+use MauticPlugin\MauticTriggerdialogBundle\Integration\TriggerdialogIntegration;
 use MauticPlugin\MauticTriggerdialogBundle\Model\TriggerCampaignModel;
+use MauticPlugin\MauticTriggerdialogBundle\Utility\SsoUtility;
 
 return [
     'name' => 'Dt. Post',
     'description' => 'Send postcards or letters via Deutsche Post TRIGGERDIALOG',
-    'version' => '1.1.0',
+    'version' => '1.2.0',
```
Comment (flossels) [inline Config/config.php:16]: Breaking changes should result in a new major version

File: Controller/TriggerCampaignController.php [hunk 2]
```diff
@@ -605,10 +605,11 @@ protected function createTriggerCampaignModifyRequest(TriggerCampaign $triggerCa
             //lock the entity
             $triggerCampaignModel->lockEntity($triggerCampaign);
         }
+        $test = $this->setFormTheme($form, self::TEMPLATES['form'], self::THEMES['variables']);
```
Comment (flossels) [inline Controller/TriggerCampaignController.php:608]: Why do you declare this? -> DRY

File: Controller/TriggerCampaignController.php [hunk 3]
```diff
@@ -605,10 +605,11 @@ protected function createTriggerCampaignModifyRequest(TriggerCampaign $triggerCa
...
 
         return $this->delegateView([
             'viewParameters' => [
-                'form' => $this->setFormTheme($form, self::TEMPLATES['form'], self::THEMES['variables']),
+                'form' => $test,
```
Comment (flossels) [inline Controller/TriggerCampaignController.php:612]: see: above

File: Entity/TriggerCampaign.php [hunk 4]
```diff
@@ -30,6 +30,16 @@ class TriggerCampaign extends FormEntity
      */
     private $id;
 
+    /**
+     * @var null|int
+     */
+    private $triggerId;
```
Comment (flossels) [inline Entity/TriggerCampaign.php:36]: Please harden this to be an int by default:

`private $triggerId = 0;`

File: Entity/TriggerCampaign.php [hunk 5]
```diff
@@ -30,6 +30,16 @@ class TriggerCampaign extends FormEntity
...
+
+    /**
+     * @var null|int
+     */
+    private $mailingId;
```
Comment (flossels) [inline Entity/TriggerCampaign.php:41]: Please harden this to be an int by default:

`private $mailingId = 0;`

File: Entity/TriggerCampaign.php [hunk 6]
```diff
@@ -81,6 +91,8 @@ public function __clone()
     public function __construct()
     {
         $this->setStartDate();
+        $this->setMailingId(0);
```
Comment (flossels) [inline Entity/TriggerCampaign.php:94]: This is not necessary when properties are int by default.

File: Entity/TriggerCampaign.php [hunk 7]
```diff
@@ -244,12 +257,20 @@ public function getVariables()
      */
     public function getVariablesAsArray()
     {
+        $variableDefDataType = json_decode('[{"id": 10,"label": "string"},{"id": 20,"label": "integer"},{"id": 30,"label": "boolean"},{"id": 40,"label": "date"},{"id": 50,"label": "image"},{"id": 60,"label": "imageurl"},{"id": 70,"label": "float"},{"id": 80,"label": "zip"},{"id": 90,"label": "countryCode"}]', true);
```
Comment (flossels) [inline Entity/TriggerCampaign.php:260]: ???

File: Entity/TriggerCampaign.php [hunk 8]
```diff
@@ -244,12 +257,20 @@ public function getVariables()
...
         $variables = [];
 
         foreach ($this->variables as $variable) {
+            $type_def = '';
```
Comment (flossels) [inline Entity/TriggerCampaign.php:264]: Please do not switch between lowercase_underscore and lowerCamelCase

File: Entity/TriggerCampaign.php [hunk 9]
```diff
@@ -300,4 +321,36 @@ public function setPrintNodeDescription($printNodeDescription)
         $this->isChanged('printNodeDescription', $printNodeDescription);
         $this->printNodeDescription = $printNodeDescription;
     }
+
+    /**
+     * @return null|int
```
Comment (flossels) [inline Entity/TriggerCampaign.php:326]: Method should return int only.

File: Entity/TriggerCampaign.php [hunk 10]
```diff
@@ -300,4 +321,36 @@ public function setPrintNodeDescription($printNodeDescription)
...
+    {
+        return $this->triggerId;
+    }
+
+    /**
+     * @param null|int $triggerId
+     */
+    public function setTriggerId($triggerId): void
+    {
+        $this->triggerId = $triggerId;
```
Comment (flossels) [inline Entity/TriggerCampaign.php:338]: Make sure to allow only integers here.

File: Entity/TriggerCampaign.php [hunk 11]
```diff
@@ -300,4 +321,36 @@ public function setPrintNodeDescription($printNodeDescription)
...
+    }
+
+    /**
+     * @return null|int
+     */
+    public function getMailingId(): int
+    {
+        return (int)$this->mailingId;
```
Comment (flossels) [inline Entity/TriggerCampaign.php:346]: Do not cast to int in the getter. Please move this into the setter.

File: Entity/TriggerCampaign.php [hunk 12]
```diff
@@ -300,4 +321,36 @@ public function setPrintNodeDescription($printNodeDescription)
...
+    /**
+     * @param null|int $triggerId
+     */
+    public function setTriggerId($triggerId): void
+    {
+        $this->triggerId = $triggerId;
+    }
+
+    /**
+     * @return null|int
```
Comment (flossels) [inline Entity/TriggerCampaign.php:342]: Method should return int only.

File: Entity/TriggerCampaign.php [hunk 13]
```diff
@@ -300,4 +321,36 @@ public function setPrintNodeDescription($printNodeDescription)
...
+    {
+        return (int)$this->mailingId;
+    }
+
+    /**
+     * @param null|int $mailingId
+     */
+    public function setMailingId($mailingId): void
+    {
+        $this->mailingId = $mailingId;
```
Comment (flossels) [inline Entity/TriggerCampaign.php:354]: Make sure to allow only integers here.

File: EventListener/CampaignSubscriber.php [hunk 14]
```diff
@@ -101,7 +101,7 @@ public function onTriggerCampaignPreSave(TriggerCampaignEvent $event): void
                 $this->getTriggerDialogService()->updateCampaign($triggerCampaign);
             }
             if (isset($changes['variables'])) {
-                $this->getTriggerDialogService()->updateCampaignVariable($triggerCampaign);
+                $this->getTriggerDialogService()->updateCampaignVariable($triggerCampaign, $changes['variables'][1]);
```
Comment (flossels) [inline EventListener/CampaignSubscriber.php:104]: Why `[1]`?

File: EventListener/CampaignSubscriber.php [hunk 15]
```diff
@@ -117,7 +117,9 @@ public function onTriggerCampaignPostSave(TriggerCampaignEvent $event): void
         $triggerCampaign = $event->getTriggerCampaign();
 
         if (isset($triggerCampaign->getChanges()['printNodeId'])) {
-            $this->getTriggerDialogService()->createCampaign($triggerCampaign);
+            $triggerCampaign = $this->getTriggerDialogService()->createCampaign($triggerCampaign);
+            $tcmr = $this->triggerCampaignModel->getRepository();
```
Comment (flossels) [inline EventListener/CampaignSubscriber.php:121]: Please use meaningful variable names or even better:

`$this->triggerCampaignModel->getRepository()->saveEntity($triggerCampaign);`

File: Form/Type/ConfigType.php [hunk 16]
```diff
@@ -62,40 +62,6 @@ public function buildForm(FormBuilderInterface $builder, array $options)
             ],
         ]);
 
-        $builder->add('triggerdialog_username', TextType::class, [
-            'label' => 'plugin.triggerdialog.form.username',
```
Comment (flossels) [inline Form/Type/ConfigType.php:66]: Please remove this label from the translations, too.

File: Form/Type/ConfigType.php [hunk 17]
```diff
@@ -62,40 +62,6 @@ public function buildForm(FormBuilderInterface $builder, array $options)
...
-            ],
-            'required' => true,
-            'attr' => [
-                'class' => 'form-control',
-                'rows' => '6',
-            ],
-        ]);
-
-        $builder->add('triggerdialog_firstName', TextType::class, [
-            'label' => 'plugin.triggerdialog.form.firstName',
```
Comment (flossels) [inline Form/Type/ConfigType.php:78]: Please remove this label from the translations, too.

File: Form/Type/ConfigType.php [hunk 18]
```diff
@@ -62,40 +62,6 @@ public function buildForm(FormBuilderInterface $builder, array $options)
...
-                'class' => 'control-label',
-            ],
-            'attr' => [
-                'class' => 'form-control',
-                'rows' => '6',
-            ],
-        ]);
-
-        $builder->add('triggerdialog_lastName', TextType::class, [
-            'label' => 'plugin.triggerdialog.form.lastName',
```
Comment (flossels) [inline Form/Type/ConfigType.php:89]: Please remove this label from the translations, too.

File: Library/vendor/firebase/php-jwt/src/JWT.php [hunk 19]
```diff
@@ -109,7 +109,7 @@ public static function decode($jwt, $key, array $allowed_algs = array())
 
         // Check the signature
         if (!static::verify("$headb64.$bodyb64", $sig, $key, $header->alg)) {
-            throw new SignatureInvalidException('Signature verification failed');
+            //throw new SignatureInvalidException('Signature verification failed');
```
Comment (flossels) [inline Library/vendor/firebase/php-jwt/src/JWT.php:112]: ???

File: Service/TriggerdialogService.php [hunk 20]
```diff
@@ -23,199 +29,341 @@ class TriggerdialogService
...
         return self::$_instance;
     }
 
+    public function setJWT()
+    {
+        $this->jwt = $this->getAuthorizationJWT();
+        try {
+            $this->jwtKeys = $array = json_decode(json_encode(JWT::decode($this->jwt, 'aKaqioatnPqwSrWWy5-9v', ['HS512'])), true);
+        } catch (\Exception $e) {
+            var_dump($e); //todo: log exception
```
Comment (flossels) [inline Service/TriggerdialogService.php:112]: ???

File: Service/TriggerdialogService.php [hunk 21]
```diff
@@ -23,199 +29,341 @@ class TriggerdialogService
...
         }
 
         return self::$_instance;
     }
 
+    public function setJWT()
+    {
+        $this->jwt = $this->getAuthorizationJWT();
+        try {
+            $this->jwtKeys = $array = json_decode(json_encode(JWT::decode($this->jwt, 'aKaqioatnPqwSrWWy5-9v', ['HS512'])), true);
```
Comment (flossels) [inline Service/TriggerdialogService.php:110]: ???

File: Service/TriggerdialogService.php [hunk 22]
```diff
@@ -23,199 +29,341 @@ class TriggerdialogService
...
      * @throws \GuzzleHttp\Exception\GuzzleException
      */
-    public function createCampaign(TriggerCampaign $triggerCampaign)
+    public function createCampaign(TriggerCampaign $triggerCampaign): TriggerCampaign
     {
         $data = $this->getCampaignData($triggerCampaign);
-        $data['variable'] = $triggerCampaign->getVariablesAsArray();
-
+        //$data['variable'] = $triggerCampaign->getVariablesAsArray();
+        $json_body = $data;
```
Comment (flossels) [inline Service/TriggerdialogService.php:157]: Please do not switch between lowercase_underscore and lowerCamelCase

File: Service/TriggerdialogService.php [hunk 23]
```diff
@@ -23,199 +29,341 @@ class TriggerdialogService
...
+            'POST',
+            '/gateway/longtermcampaigns',
+            [
+                'json' => $json_body,
+                'headers' => ['Authorization' => $this->jwt],
+            ]
+        );
 
-        if ($response->getStatusCode() !== 200) {
+        $response_body = json_decode($response->getBody()->getContents(), true);
```
Comment (flossels) [inline Service/TriggerdialogService.php:170]: Please do not switch between lowercase_underscore and lowerCamelCase

File: Service/TriggerdialogService.php [hunk 24]
```diff
@@ -23,199 +29,341 @@ class TriggerdialogService
...
 
-        $xml = new \SimpleXMLElement('<updateCampaignRequest xmlns:ns2="urn:pep-dpdhl-com:triggerdialog/campaign/v_10"></updateCampaignRequest>');
-        $this->transformData($xml, $data);
+        return $triggerCampaign;
+    }
 
-        $response = $this->client->request('POST', '/rest-mas/campaign/', ['body' => $xml->asXML()]);
+    public function setVariableDefinitions(TriggerCampaign $triggerCampaign): void
+    {
+        $json_body = [
```
Comment (flossels) [inline Service/TriggerdialogService.php:185]: Please do not switch between lowercase_underscore and lowerCamelCase.

File: Service/TriggerdialogService.php [hunk 25]
```diff
@@ -23,199 +29,341 @@ class TriggerdialogService
...
+            'customerId' => $this->jwtKeys['customerIds'][0],
```
Comment (flossels) [inline Service/TriggerdialogService.php:186]: Are you sure that index `0` exists?

File: Service/TriggerdialogService.php [hunk 26]
```diff
@@ -23,199 +29,341 @@ class TriggerdialogService
...
     }
 
-    public function addCampaignPrintNode()
+    /**
+     * @param TriggerCampaign $triggerCampaign
+     */
+    public function createMailing(TriggerCampaign $triggerCampaign): void
     {
-        // TODO: Implement later on
+        $json_body = [
```
Comment (flossels) [inline Service/TriggerdialogService.php:209]: Please do not switch between lowercase_underscore and lowerCamelCase.

File: Service/TriggerdialogService.php [hunk 27]
```diff
@@ -23,199 +29,341 @@ class TriggerdialogService
...
+        $response = $this->client->request(
+            'POST',
+            '/gateway/mailings',
+            [
+                'json' => $json_body,
+                'headers' => ['Authorization' => $this->jwt],
+            ]
+        );
+
+        $response_body = json_decode($response->getBody()->getContents(), true);
```
Comment (flossels) [inline Service/TriggerdialogService.php:223]: Please do not switch between lowercase_underscore and lowerCamelCase.

File: Service/TriggerdialogService.php [hunk 28]
```diff
@@ -23,199 +29,341 @@ class TriggerdialogService
...
      */
-    public function updateCampaignVariable(TriggerCampaign $triggerCampaign)
+    public function updateCampaign(TriggerCampaign $triggerCampaign, $state = 'active'): void
     {
-        $data = $this->getCampaignData($triggerCampaign, false);
-        $data['variable'] = $triggerCampaign->getVariablesAsArray();
-
-        $xml = new \SimpleXMLElement('<updateCampaignVariableRequest xmlns:ns2="urn:pep-dpdhl-com:triggerdialog/campaign/v_10"></updateCampaignVariableRequest>');
+        $data = $this->getCampaignData($triggerCampaign);
+        $json_body = $data;
```
Comment (flossels) [inline Service/TriggerdialogService.php:237]: Please do not switch between lowercase_underscore and lowerCamelCase.

File: Service/TriggerdialogService.php [hunk 29]
```diff
@@ -23,199 +29,341 @@ class TriggerdialogService
...
+    /**
+     * @param TriggerCampaign $triggerCampaign
+     *
+     * @param null $diff
+     * @throws RequestException
+     */
+    public function updateCampaignVariable(TriggerCampaign $triggerCampaign, $diff = null): void
+    {
+        $variables = $this->transformVariableArray($diff);
+        $json_body = [
```
Comment (flossels) [inline Service/TriggerdialogService.php:260]: Please do not switch between lowercase_underscore and lowerCamelCase.

**Review by flossels — CHANGES_REQUESTED** [review]
Please do not switch between lowercase_underscore and lowerCamelCase.

File: Service/TriggerdialogService.php (hunk 25 above)
Comment (luguenth) [reply Service/TriggerdialogService.php:184]: Yes, there's always one Customer

File: EventListener/CampaignSubscriber.php (hunk 14 above)
Comment (luguenth) [reply EventListener/CampaignSubscriber.php:104]: In [1] is always the changed new dataset

**Pull request closed without merging**